File: hw3/jiaqi_fan_hw3/jiaqi_fan_task2.scala
import java.io.PrintWriter

import org.apache.spark.{SparkConf, SparkContext}
import org.apache.spark.mllib.recommendation.{ALS, Rating}

object jiaqi_fan_task2 {



  def main(args: Array[String]): Unit = {
    val start = System.currentTimeMillis()
    val conf = new SparkConf()
      .setAppName("INF553_hw3")
      .setMaster("local[*]")
    val sc =  new SparkContext(conf)
    val train = sc.textFile(args(0))
    val test = sc.textFile(args(1))
    val case_num = args(2).toInt
//
//    val train = sc.textFile("/Users/frank/Desktop/553/hw_data/hw3/yelp_train.csv")
//    val test = sc.textFile("/Users/frank/Desktop/553/hw_data/hw3/yelp_val.csv")
//    val case_num = 2


    val train_header = train.first()
    val train_data = train.filter(x=> x!= train_header).map(x => x.split(",")).map(x => (x(0), x(1), x(2))).cache()
    val test_header = test.first()
    val test_data = test.filter(x=> x!= test_header).map(x => x.split(",")).map(x => (x(0), x(1), x(2))).cache()


    if(case_num == 1) {

      val user = train_data.map(x => x._1).union(test_data.map(x => x._1)).distinct().collect()
      val business = train_data.map(x => x._2).union(test_data.map(x => x._2)).distinct().collect()

      var user_id_map = scala.collection.mutable.Map[String, Int]()
      var business_id_map = scala.collection.mutable.Map[String, Int]()
      var id_user_map = scala.collection.mutable.Map[Int, String]()
      var id_business_map = scala.collection.mutable.Map[Int, String]()
      var num = 0
      for (i <- user) {
        user_id_map += (i -> num)
        id_user_map += (num -> i)
        num += 1
      }
      num = 0
      for (i <- business) {
        business_id_map += (i -> num)
        id_business_map += (num -> i)
        num += 1
      }

      val test = test_data.map(x => (user_id_map(x._1), business_id_map(x._2)))
      val training = train_data.map(x => (x._1, x._2, x._3) match {case (user, business, rating) => Rating( user_id_map(user), business_id_map(business),rating.toDouble )})

      val rank = 5
      val numIterations = 10
      val model = ALS.train(training, rank, numIterations, 0.2)

      val predictions = model.predict(test).map{case Rating(user, business, ratings) => ((id_user_map(user), id_business_map(business)),ratings)}

      val output = new PrintWriter(args(3))
//      val output = new PrintWriter("output.csv")
      output.write("user_id, business_id, prediction\n")
      for(ele <- predictions.collect()){
        output.write( ele._1._1 + "," + ele._1._2 + "," +ele._2 + "\n")
      }

      val rmse = test_data.map(x => ((x._1, x._2), x._3.toDouble)).join(predictions).map(x => Math.abs(x._2._1 - x._2._2)).map(x => x*x).mean()
      println(rmse)

    }


    if (case_num == 2){
        val get_value = train_data.map(x => ((x._1, x._2), x._3.toDouble)).collectAsMap()
        val user_ave_value = train_data.map(x=> (x._1, (x._3.toDouble, 1))).reduceByKey((x,y) => (x._1 + y._1, x._2 + y._2)).map(x => (x._1,x._2._1/x._2._2)).collectAsMap()

        val user_all_business = train_data.map(x=> (x._1,x._2)).groupByKey().mapValues(x=> x.toSet).collectAsMap()
        val business_all_user = train_data.map(x=> (x._2,x._1)).groupByKey().mapValues(x=> x.toSet).collectAsMap()


        val test_pair = test_data.map(x => (x._1, x._2)).collect()

        var predictions = scala.collection.mutable.ListBuffer[((String,String),Double)]()
        for (pair <- test_pair){

          if(!user_all_business.contains(pair._1)) {
                predictions += Tuple2((pair._1, pair._2), 3.0)
          }else if (!business_all_user.contains(pair._2)){
                predictions += Tuple2((pair._1, pair._2), 3.0)
          }else{
              var user_weight = scala.collection.mutable.ListBuffer[(Double,Double)]()
              for (user <- business_all_user(pair._2)){

                val a_set = user_all_business(pair._1)
                val u_set = user_all_business(user)
                val co_set = a_set.intersect(u_set)

                val ave_u = user_ave_value(user)
                val ave_a = user_ave_value(pair._1)

                var up_num = 0.0
                var de1 = 0.0
                var de2 = 0.0

                for(item <- co_set){
                  up_num += (get_value((user,item)) - ave_u) * (get_value((pair._1,item)) - ave_a)
                  de1 += (get_value((user,item)) - ave_u) * (get_value((user,item)) - ave_u)
                  de2 += (get_value((pair._1,item)) - ave_a) * (get_value((pair._1,item)) - ave_a)
                }

                val lo_num = math.sqrt(de1)* math.sqrt(de2)
                var co_rated = 0.0
                if (lo_num!=0) {
                  co_rated = up_num / lo_num
                }

                user_weight += Tuple2(co_rated, get_value((user,pair._2)) - user_ave_value(user))
              }

            var up_sum = 0.0
            var lo_sum = 0.0

            for(x <- user_weight){
              up_sum += x._1 * x._2
              lo_sum += math.abs(x._1)
            }
            var pre = user_ave_value(pair._1)
            if(lo_sum != 0)
              pre  += up_sum/lo_sum

            predictions += Tuple2((pair._1, pair._2), pre)

          }

          val output = new PrintWriter(args(3))
          //      val output = new PrintWriter("output.csv")
          output.write("user_id, business_id, prediction\n")
          for(ele <- predictions.toList){
            output.write( ele._1._1 + "," + ele._1._2 + "," +ele._2 + "\n")
          }

          val p = sc.parallelize(predictions)
          val rmse = test_data.map(x => ((x._1, x._2), x._3.toDouble)).join(p).map(x => Math.abs(x._2._1 - x._2._2)).map(x => x*x).mean()
          println(rmse)

        }
    }


    val end = System.currentTimeMillis()
    val During = (end - start) /1000
    println("Duration: " + During)

    }
}

File: hw3/jiaqi_fan_hw3/jiaqi_fan_task2.py
from pyspark import SparkContext
from pyspark.mllib.recommendation import ALS, Rating
import sys, time, math


def correct_rating(value):
    if value < 1:
        return 1.0
    elif value < 5:
        return value
    else:
        return 5.0


start = time.time()
sc = SparkContext("local[*]", "inf553_hw3")
train = sc.textFile(sys.argv[1]).cache()
test = sc.textFile(sys.argv[2]).cache()
case_id = int(sys.argv[3])


# train = sc.textFile("/Users/frank/Desktop/553/hw_data/hw3/yelp_train.csv").cache()
# test = sc.textFile("/Users/frank/Desktop/553/hw_data/hw3/yelp_val.csv")
train_header = train.first()
train_data = train.filter(lambda record: record != train_header).map(lambda record: record.split(",")).cache()
test_header = test.first()
test_data = test.filter(lambda record: record != test_header).map(lambda record: record.split(",")).cache()

if case_id == 1:
    t = test_data.map(lambda x: (x[0], x[1]))
    user_business = train_data.map(lambda x: (x[0], x[1])).union(t)
    user = user_business.map(lambda x: (x[0], 1)).reduceByKey(lambda x, y: x + y).keys()
    business = user_business.map(lambda x: (x[1], 1)).reduceByKey(lambda x, y: x + y).keys()
    user_map = {}
    id_user = {}
    business_map = {}
    id_business = {}

    for user_id, user_name in enumerate(user.collect()):
        user_map[user_name] = user_id
        id_user[user_id] = user_name

    for business_id, business_name in enumerate(business.collect()):
        business_map[business_name] = business_id
        id_business[business_id] = business_name

    test_pair = test_data.map(lambda record: (user_map[record[0]], business_map[record[1]]))
    train_rating = train_data.map(lambda record: Rating(user_map[record[0]], business_map[record[1]], float(record[2])))

    rank = 5
    numIterations = 10
    lambda_num = 0.3
    model = ALS.train(train_rating, rank, numIterations, lambda_num)

    predictions = model.predictAll(test_pair).map(lambda r: ((id_user[r[0]], id_business[r[1]]), r[2]))

    output = open(sys.argv[4], "w")
    # output = open("output.csv", "w")
    predict_result = sorted(predictions.collect())

    print("user_id, business_id, prediction", file=output)
    for ele in predict_result:
        print(ele[0][0] + "," + ele[0][1] + "," + str(correct_rating(ele[1])), file=output)

    output.close()
    rate_system = test_data.map(lambda record: ((record[0], record[1]), float(record[2]))).join(predictions).map(lambda record: abs(record[1][0]-record[1][1])).map(lambda x: x*x).mean()
    rmse = math.sqrt(rate_system)


if case_id == 2:
    user_ave_value = train_data.map(lambda record: (record[0], (float(record[2]), 1))).reduceByKey(lambda x, y: (x[0]+y[0], x[1]+y[1])).map(lambda x: (x[0], (x[1][0]/x[1][1]))).collectAsMap()
    business_ave_value = train_data.map(lambda x: (x[1], (float(x[2]), 1))).reduceByKey(lambda x, y: (x[0]+y[0], x[1]+y[1])).map(lambda x: (x[0], x[1][0]/x[1][1])).collectAsMap()
    user_value = train_data.map(lambda x: (x[0], (x[1], float(x[2]) - user_ave_value[x[0]]))).groupByKey().mapValues(dict).collectAsMap()

    business_alluser = train_data.map(lambda x: (x[1], x[0])).groupByKey().mapValues(set).collectAsMap()
    user_allbusiness = train_data.map(lambda x: (x[0], x[1])).groupByKey().mapValues(set).collectAsMap()

    test_pair = test_data.map(lambda x: (x[0], x[1])).collect()

    predictions = []
    for pairs in test_pair:
        if pairs[1] not in business_alluser:
            predictions.append((pairs[0], pairs[1], user_ave_value[pairs[0]]))
        elif pairs[0] not in user_allbusiness:
            predictions.append((pairs[0], pairs[1], business_ave_value[pairs[1]]))
        else:
            if pairs[1] in user_value[pairs[0]]:
                predictions.append((pairs[0], pairs[1], user_value[pairs[0]][pairs[1]] + user_ave_value[pairs[0]]))
            else:
                correlation = []
                for ele in business_alluser[pairs[1]]:
                    co_items = user_allbusiness[pairs[0]].intersection(user_allbusiness[ele])
                    sum_num = 0
                    de1 = 0
                    de2 = 0
                    for item in co_items:
                        sum_num += user_value[pairs[0]][item] * user_value[ele][item]
                        de1 += user_value[pairs[0]][item] * user_value[pairs[0]][item]
                        de2 += user_value[ele][item] * user_value[ele][item]
                    co_num = 0
                    if de1 != 0 and de2 != 0:
                        co_num = sum_num / math.sqrt(de1 * de2)

                    correlation.append((co_num, user_value[ele][pairs[1]]))

                correlation = sorted(correlation)
                N = 30
                if len(correlation) > N:
                    for i in range(len(correlation) -N):
                        del correlation[0]

                upper_sum = 0
                lower_sum = 0
                for num in correlation:
                    upper_sum += float(num[0]) * num[1]
                    lower_sum += abs(float(num[0]))

                pred = user_ave_value[pairs[0]]
                if lower_sum != 0:
                    pred = upper_sum/lower_sum + user_ave_value[pairs[0]]

                predictions.append((pairs[0], pairs[1], correct_rating(pred)))

    output = open(sys.argv[4], "w")
    # output = open("output.csv", "w")
    print("user_id, business_id, prediction", file=output)
    for ele in predictions:
        print(ele[0], "," + ele[1], "," + str(ele[2]), file=output)
    output.close()

    ddddd = sc.parallelize(predictions).map(lambda x: ((x[0], x[1]), float(x[2])))
    rate_system = test_data.map(lambda record: ((record[0], record[1]), float(record[2]))).join(ddddd).map(lambda record: abs(record[1][0]-record[1][1])).map(lambda x: x*x).mean()
    rmse = math.sqrt(rate_system)


if case_id == 3:
    business_alluser = train_data.map(lambda x: (x[1], x[0])).groupByKey().mapValues(set).collectAsMap()
    user_allbusiness = train_data.map(lambda x: (x[0], x[1])).groupByKey().mapValues(set).collectAsMap()
    get_value = train_data.map(lambda x: ((x[0], x[1]), float(x[2]))).collectAsMap()

    def cold_start(x):
        if x[0] not in user_allbusiness or x[1] not in business_alluser:
            return ()
        else:
            return (x[0] ,x[1])


    def item_weight(a, b, c):
        set_a = business_alluser[a]
        set_b = business_alluser[b]
        coreted_item = (set_a & set_b) - {c}
        if len(coreted_item) == 0:
            return 0
        else:
            a_list = []
            b_list = []
            for x in coreted_item:
                a_list.append(get_value[(x, a)])
                b_list.append(get_value[(x, b)])
            A_ave = sum(a_list) / len(a_list)
            B_ave = sum(b_list) / len(b_list)

            numerator = 0
            A_square = 0
            B_square = 0
            for i in range(len(coreted_item)):
                numerator += (a_list[i] - A_ave) * (b_list[i] - B_ave)
                A_square += (a_list[i] - A_ave) ** 2
                B_square += (b_list[i] - B_ave) ** 2

            denominator = math.sqrt(A_square) * math.sqrt(B_square)
            if denominator == 0:
                return None
            return numerator / denominator


    def item(x):
        sum_all = 0
        count = 0
        for item in business_alluser[x[1]]:
            if item == x[0]:
                continue
            sum_all += get_value[(item, x[1])]
            count += 1

        ave = 0.0
        if count != 0:
            ave = float(sum_all / count)
        co_rated = []
        for ele in user_allbusiness[x[0]]:
            if ele == x[1]:
                continue
            co = item_weight(x[1], ele, x[0])
            if co == None:
                return (x, ave)
            co_rated.append([x[0], ele, co])
        co_rated.sort(key=lambda x: x[2], reverse=True)

        length = 5
        if length > len(co_rated):
            length = len(co_rated)

        up_list = []
        for i in range(length):
            up_list.append(get_value[(co_rated[i][0], co_rated[i][1])] * co_rated[i][2])
        up_num = sum(up_list)
        lo_list = []
        for i in range(length):
            lo_list.append(abs(co_rated[i][2]))
        lo_num = sum(lo_list)

        pre_value = ave
        if lo_num != 0:
            pre_value = ave + up_num / lo_num
        return (x, pre_value)


    process_data = test_data.map(cold_start).filter(lambda x: len(x) != 0)
    not_exist_data = test_data.map(lambda x: ((x[0], x[1]), 3)).subtractByKey(process_data.map(lambda x: (x, 1)))
    predictions = process_data.map(lambda x: item(x))
    result = sc.union([predictions, not_exist_data])

    output = open(sys.argv[4], "w")
    print("user_id, business_id, prediction", file=output)
    for ele in result.collect():
        print(ele[0][0], "," + ele[0][1], "," + str(ele[1]), file=output)
    output.close()

    rate_system = test_data.map(lambda x: ((x[0], x[1]), float(x[2]))).join(result).map(lambda record: abs(record[1][0]-record[1][1])).map(lambda x: x*x).mean()
    rmse = math.sqrt(rate_system)


duration = time.time() - start
print("Duration: ", duration)
print("RMSE: ", str(rmse))


